packages/profile-sync-controller/src/shared/encryption/encryption.ts

```diff
@@ -250,33 +258,85 @@ class EncryptorDecryptor {
       };
     }
 
+    // Create a unique cache key for this KDF operation
     const newSalt = salt ?? SHARED_SALT;
+    const cacheKey = this.#createKdfCacheKey(
+      hashedPassword,
+      o,
+      newSalt,
+      nativeScryptCrypto,
+    );
+
+    // Check if there's already an ongoing KDF operation with the same parameters
+    const existingPromise = this.#kdfPromiseCache.get(cacheKey);
+    if (existingPromise) {
+      return existingPromise;
+    }
+
+    // Create and cache the promise for the KDF operation
+    const kdfPromise = this.#performKdfOperation(
+      password,
+      o,
+      newSalt,
+      hashedPassword,
+      nativeScryptCrypto,
+    );
+
+    // Cache the promise and set up cleanup
+    this.#kdfPromiseCache.set(cacheKey, kdfPromise);
```

Comment:
Although it's unlikely if promises never resolve or reject the map grows indefinitely, doesn't it? 

Do we need to manage overflow like
```suggestion
  if (this.#kdfPromiseCache.size >= this.#MAX_CACHE_SIZE) {
    this.#kdfPromiseCache.clear();
  }

    this.#kdfPromiseCache.set(cacheKey, kdfPromise);
```
or does `set` already take care of it internally or smth like that?

Reply:
Done with [309edae](https://github.com/MetaMask/core/pull/6736/commits/309edae4a4b7dd9de542ff85bc1c58dd46f9afb2)